storage/bucket.go

```diff
@@ -0,0 +1,347 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/coreos/mantle/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/mantle/Godeps/_workspace/src/google.golang.org/api/googleapi"
+	"github.com/coreos/mantle/Godeps/_workspace/src/google.golang.org/api/storage/v1"
+)
+
+var (
+	UnknownScheme = errors.New("storage: URL missing gs:// scheme")
+	UnknownBucket = errors.New("storage: URL missing bucket name")
+)
+
+type Bucket struct {
+	service *storage.Service
+	name    string
+	prefix  string
+
+	mu      sync.RWMutex
+	objects map[string]*storage.Object
+
+	// writeAlways enables overwriting of objects that appear up-to-date
+	writeAlways bool
+	// writeDryRun blocks any changes, merely logging them instead
+	writeDryRun bool
+}
+
+func NewBucket(client *http.Client, bucketURL string) (*Bucket, error) {
+	service, err := storage.New(client)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedURL, err := url.Parse(bucketURL)
+	if err != nil {
+		return nil, err
+	}
+	if parsedURL.Scheme != "gs" {
+		return nil, UnknownScheme
+	}
+	if parsedURL.Host == "" {
+		return nil, UnknownBucket
+	}
+
+	// unsure if this is behavior is what we want or not
+	if parsedURL.Path != "" && !strings.HasSuffix(parsedURL.Path, "/") {
+		parsedURL.Path += "/"
+	}
+
+	return &Bucket{
+		service: service,
+		name:    parsedURL.Host,
+		prefix:  strings.TrimPrefix(parsedURL.Path, "/"),
+		objects: make(map[string]*storage.Object),
+	}, nil
+}
+
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+func (b *Bucket) Prefix() string {
+	return b.prefix
+}
+
+func (b *Bucket) URL() *url.URL {
+	return &url.URL{Scheme: "gs", Host: b.name, Path: b.prefix}
+}
+
+func (b *Bucket) WriteAlways(always bool) {
+	b.writeAlways = always
+}
+
+func (b *Bucket) WriteDryRun(dryrun bool) {
+	b.writeDryRun = dryrun
+}
+
+func (b *Bucket) Object(objName string) *storage.Object {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.objects[objName]
+}
+
+// Check of an object name's directory is the specified prefix.
+func matchPrefix(path, prefix string) bool {
+	if prefix == "" {
+		return true
+	}
+	i := strings.LastIndexByte(path, '/')
+	return path[:i+1] == prefix
+}
+
+func (b *Bucket) Objects(prefix string) map[string]*storage.Object {
+	objs := make(map[string]*storage.Object)
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	for objName, obj := range b.objects {
+		if matchPrefix(objName, prefix) {
+			objs[objName] = obj
+		}
+	}
+	return objs
+}
+
+func (b *Bucket) Prefixes() []string {
+	prefixmap := make(map[string]bool)
```

Comment:
Ok, sure, I can use struct{} and _, ok = but I think that looks worse and cluttered. I suppose worse and cluttered is idiomatic Go so I can do that :-P